akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamSourceStage.scala

```diff
@@ -41,159 +41,57 @@ final private[stream] class OutputStreamSourceStage(writeTimeout: FiniteDuration
-
-      private val downstreamCallback: AsyncCallback[Try[ByteString]] =
-        getAsyncCallback {
-          case Success(elem) ⇒ onPush(elem)
-          case Failure(ex)   ⇒ failStage(ex)
-        }
-
-      private val upstreamCallback: AsyncCallback[(AdapterToStageMessage, Promise[Unit])] =
+      val upstreamCallback: AsyncCallback[AdapterToStageMessage] =
         getAsyncCallback(onAsyncMessage)
 
-      def wakeUp(msg: AdapterToStageMessage): Future[Unit] = {
-        val p = Promise[Unit]()
-        upstreamCallback.invoke((msg, p))
-        p.future
-      }
-
-      private def onAsyncMessage(event: (AdapterToStageMessage, Promise[Unit])): Unit =
-        event._1 match {
-          case Flush ⇒
-            flush = Some(event._2)
-            sendResponseIfNeed()
+      private def onAsyncMessage(event: AdapterToStageMessage): Unit = {
+        event match {
+          case Send(data) ⇒
+            emit(out, data)
           case Close ⇒
-            close = Some(event._2)
-            sendResponseIfNeed()
+            completeStage()
         }
-
-      private def unblockUpstream(): Boolean =
-        flush match {
-          case Some(p) ⇒
-            p.complete(Success(()))
-            flush = None
-            true
-          case None ⇒ close match {
-            case Some(p) ⇒
-              downstreamStatus.set(Canceled)
-              p.complete(Success(()))
-              close = None
-              completeStage()
-              true
-            case None ⇒ false
-          }
-        }
-
-      private def sendResponseIfNeed(): Unit =
-        if (downstreamStatus.get() == Canceled || dataQueue.isEmpty) unblockUpstream()
-
-      private def onPush(data: ByteString): Unit =
-        if (downstreamStatus.get() == Ok) {
-          push(out, data)
-          sendResponseIfNeed()
-        }
-
-      override def preStart(): Unit = {
-        // this stage is running on the blocking IO dispatcher by default, but we also want to schedule futures
-        // that are blocking, so we need to look it up
-        val actorMat = ActorMaterializerHelper.downcast(materializer)
-        dispatcher = actorMat.system.dispatchers.lookup(actorMat.settings.blockingIoDispatcher)
       }
 
       setHandler(out, new OutHandler {
         override def onPull(): Unit = {
-          implicit val ec = dispatcher
-          Future {
-            val currentThread = Thread.currentThread()
-            // keep track of the thread for postStop interrupt
-            blockingThreadRef.compareAndSet(null, currentThread)
-            try {
-              dataQueue.take()
-            } catch {
-              case _: InterruptedException ⇒
-                Thread.interrupted()
-                ByteString.empty
-            } finally {
-              blockingThreadRef.compareAndSet(currentThread, null);
-            }
-          }.onComplete(downstreamCallback.invoke)
+          semaphore.release()
```

Comment:
Does that really work correctly? `onPull` will only be called if the `emit` queue is empty, so it seems like the semaphore would only switch between 0 and 1. Or do I miss something?

Reply:
Excellent point!